### PR TITLE
Make sure policies don't mutate the batch

### PR DESCRIPTION
**Pull request by alexander-soare**
## What this does

Adds shallow copies in relevant locations to make sure the batch is not mutated.

## How it was tested

CI

### CODE REVIEW

File: lerobot/common/policies/act/modeling_act.py [hunk 1]
```diff
@@ -128,6 +129,7 @@ def forward(self, batch: dict[str, Tensor]) -> dict[str, Tensor]:
         """Run the batch through the model and compute the loss for training or validation."""
         batch = self.normalize_inputs(batch)
         if len(self.expected_image_keys) > 0:
+            batch = dict(batch)  # shallow copy so that adding a key doesn't modify the original
```
Comment (alexander-soare) [inline lerobot/common/policies/act/modeling_act.py:132]: FYI, although this shallow copy already happens in `self.normalize_inputs` I think it's safer to add it here too rather than relying on an implicit guarantee.

File: lerobot/common/policies/act/modeling_act.py [hunk 2]
```diff
@@ -467,10 +469,9 @@ def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, tuple[Tensor, Tenso
         if self.use_images:
             all_cam_features = []
             all_cam_pos_embeds = []
-            images = batch["observation.images"]
 
-            for cam_index in range(images.shape[-4]):
-                cam_features = self.backbone(images[:, cam_index])["feature_map"]
+            for cam_index in range(batch["observation.images"].shape[-4]):
```
Comment (alexander-soare) [inline lerobot/common/policies/act/modeling_act.py:473]: FYI, not relevant to this PR, but I just noticed this while trying to track changes to batch. I think it's not necessary to pollute the namespace with `images = batch["observation.images"]`

Comment (Cadene) [reply lerobot/common/policies/act/modeling_act.py:473]: Unless there is a good reason that I am missing, I would prefer to never write anything in `batch` (e.g. `batch["observation.images"] = torch.stack(...)`)

File: tests/test_policies.py [hunk 3]
```diff
@@ -161,8 +162,13 @@ def test_policy(env_name, policy_name, extra_overrides):
...
         batch[key] = batch[key].to(DEVICE, non_blocking=True)
 
-    # Test updating the policy
+    # Test updating the policy (and test that it does not mutate the batch)
+    batch_ = deepcopy(batch)
     policy.forward(batch)
+    assert set(batch) == set(batch_), "Batch keys are not the same after a forward pass."
+    assert all(
+        torch.equal(batch[k], batch_[k]) for k in batch
+    ), "Batch values are not the same after a forward pass."
```
Comment (Cadene) [inline tests/test_policies.py:171]: nice! Could we add the same test for backward ?

I am wondering if we could add the same test for torch.compile when it's ready
cc @marinabar

**Review by Cadene — COMMENTED** [review]
Left some comments. Sorry for the disagreement ^^
Happy to do a second review.

File: tests/test_policies.py (hunk 3 above)
Comment (alexander-soare) [reply tests/test_policies.py:171]: Sorry but how could the backward pass mutate the batch?

File: lerobot/common/policies/act/modeling_act.py (hunk 1 above)
Comment (alexander-soare) [reply lerobot/common/policies/act/modeling_act.py:132]: Pretty sure since it's a shallow copy the tensors are not touched. I'm not sure about torch.compile as I haven't used it, but I think we'll probably have a lot to look at regarding that.

File: lerobot/common/policies/act/modeling_act.py (hunk 2 above)
Comment (alexander-soare) [reply lerobot/common/policies/act/modeling_act.py:473]: I'm not sure I understand this. Once it's been copied in the current scope, we should be able to do anything with it (apart from mutating the values in-place as they are still not copies).

File: lerobot/common/policies/act/modeling_act.py (hunk 1 above)
Comment (Cadene) [reply lerobot/common/policies/act/modeling_act.py:132]: sounds good

File: tests/test_policies.py (hunk 3 above)
Comment (Cadene) [reply tests/test_policies.py:171]: You are right

+ we decided to not work on torch.compile yet

**Review by Cadene — APPROVED** [review]
LGTM
